### PR TITLE
Add exceptions to @future_safe; deprecate FutureResultE

### PR DESCRIPTION
**Pull request by RomanMIzulin**
# I have added exceptions to @future_safe; removed FutureResultE

## Checklist

- [x ] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x ] I have created at least one test case for the changes I have made
- [x ] I have updated the documentation for the changes I have made
- [x ] I have added my changes to the `CHANGELOG.md`

### CODE REVIEW

**Comment by RomanMIzulin** [comment]
not sure about quality of PR and deleting removed FutureResultE, but project is in beta, kinda breaking changes are allowed I guess

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,8 @@ See [0Ver](https://0ver.org/).
 
 ### Features
 
+- Add picky exceptions to `future_safe` decorator like `safe` has.
+- Remove FutureResultE
```
Comment (sobolevn) [inline CHANGELOG.md:14]: Please, don't remove it just yet. Let's first provide an alternative. Moreover, maintaining an alias is simple enough.

File: tests/test_future/test_future_result/test_future_result_decorator.py [hunk 2]
```diff
@@ -25,3 +38,26 @@ async def test_future_safe_decorator_failure():
...
+    assert isinstance(await future_instance, IOFailure)
+
+
+@pytest.mark.anyio
+@pytest.mark.xfail(raises=AssertionError)
+async def test_future_safe_decorator_w_unexpected_error():
+    """Ensure that coroutine marked with ``@future_safe``."""
+    future_instance = _coro_three('0')
+
+    await future_instance
```
Comment (sobolevn) [inline tests/test_future/test_future_result/test_future_result_decorator.py:63]:
```suggestion
    await _coro_three('0')
```

File: tests/test_future/test_future_result/test_future_result_decorator.py [hunk 3]
```diff
@@ -25,3 +38,26 @@ async def test_future_safe_decorator_failure():
...
     assert isinstance(await future_instance, IOFailure)
+
+
+@pytest.mark.anyio
+async def test_future_safe_decorator_w_expected_error():
+    """Ensure that coroutine marked with ``@future_safe``."""
+    future_instance = _coro_two(0)
+
+    assert isinstance(future_instance, FutureResult)
+    assert isinstance(await future_instance, IOFailure)
```
Comment (sobolevn) [inline tests/test_future/test_future_result/test_future_result_decorator.py:49]: We need to check that `IOFailure` has an expected error.

File: typesafety/test_future/test_future_result_container/test_future_safe_decorator.yml [hunk 4]
```diff
@@ -25,3 +25,33 @@
...
+    from typing import Optional
+    from returns.future import future_safe
+
+    @future_safe((ValueError,))
+    async def test(
+         first: int, second: Optional[str] = None, *, kw: bool = True,
+    ) -> int:
+        return 1
+
+    reveal_type(test)  # N: Revealed type is "def (first: builtins.int, second: Union[builtins.str, None] =, *, kw: builtins.bool =) -> returns.future.FutureResult[builtins.int, builtins.Exception]"
```
Comment (sobolevn) [inline typesafety/test_future/test_future_result_container/test_future_safe_decorator.yml:42]: It should be:

```suggestion
    reveal_type(test)  # N: Revealed type is "def (first: builtins.int, second: Union[builtins.str, None] =, *, kw: builtins.bool =) -> returns.future.FutureResult[builtins.int, builtins.ValueError]"
```

**Comment by sobolevn** [comment]
I will help you with that! Thanks!

**Comment by sobolevn** [comment]
Thanks a lot!

**Pull request merged**